test/jdk/sun/security/lib/cacerts/VerifyCACerts.java

```diff
@@ -385,11 +385,26 @@ public static void main(String[] args) throws Exception {
             } catch (CertificateExpiredException cee) {
                 if (!EXPIRY_EXC_ENTRIES.contains(alias)) {
                     atLeastOneFailed = true;
-                    System.err.println("ERROR: cert is expired but not in EXPIRY_EXC_ENTRIES");
+                    System.err.println("ERROR: " + alias + " cert is expired but not in EXPIRY_EXC_ENTRIES");
                 }
             } catch (CertificateNotYetValidException cne) {
                 atLeastOneFailed = true;
-                System.err.println("ERROR: cert is not yet valid");
+                System.err.println("ERROR: " + alias + " cert is not yet valid");
+            }
+
+            // Is it CA certificate?
+            if (cert.getBasicConstraints() == -1){
+                // "VeriSign Class 3 Public Primary Certification Authority - G3" is X.509 version 1 certificate
+                if (!"verisignclass3g3ca [jdk]".equals(alias)) {
+                    atLeastOneFailed = true;
+                    System.err.println("ERROR: " + alias + " BasicConstraints is missing or doesn't specify CA:true");
+                }
+            }
+            boolean[] keyUsageBits = cert.getKeyUsage();
+            // check that the KeyUsage extension, if included, asserts the keyCertSign bit
+            if (keyUsageBits != null && !keyUsageBits[5]){
```

Comment:
I would also treat no key usage extension as an error (except for the V1 root). I would also make sure it sets the cRLSign bit.